adapters/repos/db/vector/hnsw/index.go

```diff
@@ -744,3 +717,97 @@ func IsHNSWIndex(index any) bool {
 	_, ok := index.(*hnsw)
 	return ok
 }
+
+type pair struct {
+	nodeId uint64
+	level  int
+}
+
+func (h *hnsw) calculateUnreachablePoints() []uint64 {
+	h.RLock()
+	defer h.RUnlock()
+
+	visitedPairs := make(map[pair]bool)
+	candidateList := []pair{{h.entryPointID, h.currentMaximumLayer}}
+
+	for len(candidateList) > 0 {
+		currentNode := candidateList[len(candidateList)-1]
+		candidateList = candidateList[:len(candidateList)-1]
+		if !visitedPairs[currentNode] {
+			visitedPairs[currentNode] = true
+			neighbors := h.nodes[currentNode.nodeId].connectionsAtLowerLevelsNoLock(currentNode.level, visitedPairs)
```

Comment:
You need to lock the vertex whenever you access it:

```suggestion
			n := h.nodes[currentNode.nodeId]
			n.Lock()
			neighbors := n.connectionsAtLowerLevelsNoLock(currentNode.level, visitedPairs)
			n.Unlock()
```